tests/test_torch_zhen.py

```diff
@@ -0,0 +1,489 @@
+        else:
+            # this is a homogeneous ZHEN. Convert it to heterogeneous ZHEN
+            # pyre-ignore
+            token_mixer_list = [token_mixer_list] * num_layers
+
+        for i in range(num_layers):
+            layer = ZHENLayer(
+                layer_index=i,
+                emb_dim=emb_dim,
+                # pyre-ignore[6]
+                token_mixer_list=token_mixer_list[i],
+                previous_n_embs=(
+                    original_emb_num
+                    if i == 0
+                    # pyre-ignore[6]
+                    else len(token_mixer_list[i-1]) * output_emb_per_ensemble_module
+                ),
+                previous_input_embs=(
+                    original_emb_num 
+                    if i <= 1
+                    # pyre-ignore[6]
+                    else len(token_mixer_list[i-2]) * output_emb_per_ensemble_module
+                ),
+                output_embs_per_mixer=output_emb_per_ensemble_module,
+                original_n_embs=original_emb_num, 
+            )
+            self.layers.append(layer)
+
+    def forward(
+        self,
+        input: torch.Tensor,
+        skip_connection: Optional[torch.Tensor] = None,
+    ):
+        skip_connection = None  # previous layer's input
+        for layer in self.layers:
+            input, skip_connection = layer(skip_connection, input)
+
+        output = input.view(input.shape[0], -1)
+        return output
+
+    def get_dense_params(self) -> List[nn.Parameter]:
+        return list(self.parameters())
+
+
+def weight_init_func(pt_module, name_map, params, bufs):
+    # for k, m in pt_module.named_modules():
+    #     if isinstance(m, torch.nn.Linear):
+    #         params[name_map[f"{k}.weight"]] = torch.nn.init.xavier_uniform(params[name_map[f"{k}.weight"]])
+    #         params[name_map[f"{k}.bias"]] = torch.nn.init.normal(params[name_map[f"{k}.bias"]], std=1e-6)
+    return params, bufs
+
+
+class TorchZHENTest(unittest.TestCase):
+    def test_zhen_homogeneous(self):
+        B=64  # 59  # made multiples of 8
+        F=37
+        D=64
+        LAYERS=5
+        OUTPUT_PER_ENSEMBLE=48  # 50  # made multiples of 8
+        TOKENS=[TokenMixer.ATTENTION, TokenMixer.LINEAR, TokenMixer.ATTENTION, TokenMixer.CONVOLUTION, TokenMixer.DOT]
+
+        # `meta_init` allows a PyTorch model to be created with shape-only tensors as weights.
+        pt_module_gen = lambda: ZHENCollection(LAYERS, D, TOKENS, F, OUTPUT_PER_ENSEMBLE)
+
+        dataloader = [
+            (torch.empty(B, D, F), torch.empty(B, 15360)),
+            (torch.empty(B, D, F), torch.empty(B, 15360)),
+        ]
+        loss_func = lambda *args, **kwargs: torch.nn.functional.mse_loss(*args, **kwargs)
+        optim_gen = torchoptim.adam(lr=1e-3)
+        parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=num_micro_batches)
```

Comment:
num_micro_batches is not defined, maybe:
`parallel_method = alpa.PipeshardParallel(stage_mode="auto", num_micro_batches=1)`